File: 1005-maximize-sum-of-array-after-k-negations/1005-maximize-sum-of-array-after-k-negations.java
class Solution {
    public int largestSumAfterKNegations(int[] nums, int k) {
        PriorityQueue<Integer> minHeap = new PriorityQueue<>();
        int maxSum = 0;
        for(int num : nums)
            minHeap.add(num);
        while(k-- > 0) {
            int num = minHeap.poll();
            minHeap.add(-1 * num);
        }
        while(!minHeap.isEmpty())
            maxSum += minHeap.poll();
        return maxSum;
    }
}

File: 1005-maximize-sum-of-array-after-k-negations/1005-maximize-sum-of-array-after-k-negations.cpp
class Solution {
public:
    // tc - O(nlogn), sc - O(n)
    int largestSumAfterKNegations(vector<int>& nums, int k) {
        int maxSum = 0;
        priority_queue<int, vector<int>, greater<int>> minHeap;
        for(int num : nums) 
            minHeap.push(num);
        // We either take out -ves and make them +ves or we'll negate the lowest +ves possible
        while(k--) {
            int num = minHeap.top();
            minHeap.pop();
            minHeap.push(-1 * num);
        }
        while(!minHeap.empty()) {
            maxSum += minHeap.top();
            minHeap.pop();
        }
        return maxSum;
    }
};